pkg/frontend/querymiddleware/shard_active_series_response_decoder.go

```diff
@@ -0,0 +1,439 @@
+	New: func() any {
+		return &shardActiveSeriesResponseDecoder{
+			br:  bufio.NewReaderSize(nil, 4096),
+			out: make([]byte, 0, shardActiveSeriesDecoderMaxBuffSize),
+		}
+	},
+}
+
+func borrowShardActiveSeriesResponseDecoder(ctx context.Context, rc io.ReadCloser) *shardActiveSeriesResponseDecoder {
+	d := shardActiveSeriesResponseDecoderPool.Get().(*shardActiveSeriesResponseDecoder)
+	d.reset(ctx, rc)
+	return d
+}
+
+func reuseShardActiveSeriesResponseDecoder(d *shardActiveSeriesResponseDecoder) {
+	d.reset(context.Background(), nil)
+	if cap(d.out) > shardActiveSeriesDecoderMaxBuffSize {
+		return
+	}
+	shardActiveSeriesResponseDecoderPool.Put(d)
+}
+
+type shardActiveSeriesResponseDecoder struct {
+	ctx            context.Context
+	rc             io.ReadCloser
+	br             *bufio.Reader
+	out            []byte
+	readBytesCount int
+	foundDataValue bool
+	err            error
+}
+
+func (d *shardActiveSeriesResponseDecoder) reset(ctx context.Context, rc io.ReadCloser) {
+	d.ctx = ctx
+	d.rc = rc
+	d.br.Reset(rc)
+	d.out = d.out[:0]
+	d.readBytesCount = 0
+	d.foundDataValue = false
+	d.err = nil
+}
+
+func (d *shardActiveSeriesResponseDecoder) stickError(err error) {
+	if d.err != nil {
+		return
+	}
+	d.err = err
+}
+
+func (d *shardActiveSeriesResponseDecoder) close() {
+	_, _ = io.Copy(io.Discard, d.rc)
+	_ = d.rc.Close()
+}
+
+func (d *shardActiveSeriesResponseDecoder) hasDataValue() bool {
+	return d.foundDataValue && len(d.out) > 0
+}
+
+func (d *shardActiveSeriesResponseDecoder) dataValue() string {
+	if !d.foundDataValue {
+		return ""
+	}
+	return unsafe.String(unsafe.SliceData(d.out), len(d.out))
+}
+
+func (d *shardActiveSeriesResponseDecoder) decode() error {
+	if d.foundDataValue {
+		return nil // already decoded
+	}
+
+	c := d.nextToken()
+	if d.err != nil {
+		return d.err
+	}
+	switch c {
+	case '{':
+		for d.err == nil {
+			k := d.readString()
+			switch k {
+			case "data":
+				d.readDataValue()
+				if d.err != nil {
+					return d.err
+				}
+				return d.ctx.Err()
+
+			case "error":
+				d.readErrorValue()
+				if d.err != nil {
+					return d.err
+				}
+				return d.ctx.Err()
+
+			default:
+				d.skipValue()
+			}
+
+			c = d.nextToken()
+			if c == ',' {
+				continue
+			} else if c == '}' {
+				break
+			}
+		}
+
+	default:
+		return fmt.Errorf("decode: expected {, found %c", c)
+	}
+	return errors.New("expected data field at top level")
+}
+
+func (d *shardActiveSeriesResponseDecoder) readDataValue() {
+	if c := d.nextToken(); c != ':' {
+		d.stickError(fmt.Errorf("readDataValue: expected :, found %c", c))
+		return
+	}
+	c := d.nextToken()
+	switch c {
+	case '[':
+		d.out = d.out[:0]
+
+		inner := 1
+
+		c := d.readByte()
+		for d.err == nil {
+			switch c {
+			case '[':
+				inner++
+			case ']':
+				inner--
+				if inner == 0 {
+					d.foundDataValue = true
+					return
+				}
+			default:
+				d.out = append(d.out, c)
+			}
+			c = d.readByte()
+		}
```

Comment:
If I understand correctly this reads the entire value contained in this response's data field into `d.out`? If yes I think we need to revisit this, because this would lead to the same kind of OOMs that we had before response streaming if I'm not mistaken 🤔

Reply:
Sure, during implementation I didn't realize that these data fields could end up being several megabytes in length. 🤦‍♂️  

In that case, I think it would make all the sense to  send `data` array content through a channel instead, to avoid the use of an intermediate buffer. wdyt?

Reply:
Yes, the `data` field of the responses can be several hundred MBs large. The other implementation sends each item of the `data` array separately once it's parsed (see [here](https://github.com/grafana/mimir/blob/main/pkg/frontend/querymiddleware/shard_active_series.go#L335)), so I think it makes sense to do something equivalent in this approach as well.